### PR TITLE
Add support for fp16<->int, bfloat16<->int conversions

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This PR fixes an error in torchbench:Super_SloMo benchmark and a few others.
```
RuntimeError: Illegal Cast value from  DataType: int64_t to DataType: __half
```

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 1]
```diff
@@ -367,6 +367,15 @@ Val* castOp(DataType dtype, Val* v1) {
     return set(v1);
   }
 
+  // There are no direct cast for some combinations of dtypes
+  // for example integers to fp16, bfloat16 and back
+  // so we need to try doing a conversion to float first.
+  if (dtype != DataType::Float &&
+      cast_func_str(std::make_pair(v1->getDataType().value(), dtype)) ==
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/arith.cpp:374]: This feels a little bit not ideal. Lol.

Should we actually add spot support for cast functions instead? cc'ing @kevinstephano

**Review by jjsjann123 — APPROVED** [review]
I wonder if we want this blank support for cast, I'd somewhat prefer us to add cast specific for fp16/bf16 to int.

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (IvanYashchuk) [reply torch/csrc/jit/codegen/cuda/arith.cpp:374]: I was hesitant to modify `cast_func_str` as it appears also in codegen.cpp, ir_iostream.cpp. One hop through the float dtype seemed like a clear choice to me.

I can also move the changes to `cast_func_str` if that's preferred.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/arith.cpp:374]: You'll need to update `supported_casts2string` instead and also plumb some inline ptx to the runtime files for conversion.

Totally agree that the changes in this PR is more general and cleaner. I'm just not sure if we DO want this hammer :laughing: Already approved, so feel free to merge as-is. I'm just seeking some second opinion from Kevin, who has much better instincts than I do on how cuda code translates to perf.

**Comment by IvanYashchuk** [comment]
Discussed offline:
There should be PTX instructions for casting and we should prefer to use that instead of double cast.

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/arith.cpp:374]: @kevinstephano please see ^^ and comment

Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/arith.cpp:374]: We talked about this last week where we were interested in looking for a PTX version that does this officially.  The model where this happens is a little weird in the first place as it looks like some generic editing for autocast in torchbench was done that might have not been totally desirable but it isn't crazy that a real user might do this.

**Review by jjsjann123 — CHANGES_REQUESTED** [review]
I'm removing my stamp to avoid accidental merge, since we discussed that this should be implemented with a precise conversion via inline-PTX.

Feel free to yell at me if the plan changes and we'd like to merge this as a temporary WAR and leaving inline-PTX cast in a future PR.

**Comment by csarofeen** [comment]
Closing as https://github.com/csarofeen/pytorch/pull/2084 was merged in.

**Pull request closed without merging**